python/ql/src/Security/CWE-116/BadTagFilter.ql

```diff
@@ -0,0 +1,19 @@
+/**
+ * @name Bad HTML filtering regexp
+ * @description Matching HTML tags using regular expressions is hard to do right, and can easily lead to security issues.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.8
+ * @precision high
+ * @id py/bad-tag-filter
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-116
+ *       external/cwe/cwe-020
+ */
+
+import semmle.python.security.BadTagFilterQuery
+
+from HTMLMatchingRegExp regexp, string msg
+where msg = min(string m | isBadRegexpFilter(regexp, m) | m order by m.length(), m) // there might be multiple, we arbitrarily pick the shortest one
```

Comment:
why not show all problems with a single regexp all at once? Couldn't we easily end up in a situation where someone does `n` iterations of fixing up a regexp to pass our tests? (waiting `n` times for our CodeQL actions to complete analyzing the code)

Reply:
I very much prefer doing some sort of de-duplication. 

E.g. a regexp like `<script.*?>.*?<\/script>` would cause 5 alerts without any de-duplication, and the extra messages are not very helpful. 
Here are the messages (with the current query implementation): 
```
This regular expression does not match script end tags like </script >
This regular expression does not match script end tags like </script foo="bar">
This regular expression does not match script end tags like </script\\t\\n bar>
This regular expression matches <script>...</script>, but not <script >...\\n</script>
This regular expression matches <script></script>, but not <script \\n></script>
```

I could group the messages into categories, and show max 1 message from each category?

Reply:
Ahh yes, that would probably also end up opening up 5 different alerts with Code Scanning (if my understanding is correct), which seems a bit silly.

One solution could be to add ` (and 4 other edge-cases)` (and leave out that part if there are 0 other edge-cases). So the alert would look like:

```
This regular expression does not match script end tags like </script > (and 4 other cases)
```

I'm not sure I'm a big fan of this formulation either, but at least it would let people know that there is more work to be done :shrug: 

I'm probably only feeling this is 2/10 important, so I would also be OK with leaving it as is. Just saw it as an opportunity to remove a small annoyance end-users might have.